File: lifegame.hs
import Data.Array

data Lifegame = Lifegame { size :: Int
                         , board :: [(Int, Int)]
                         }

newGeneration :: Lifegame -> Lifegame
newGeneration (Lifegame size board) = (Lifegame size [(x,y) |x <- [0..size-1], y <- [0..size-1], isAlive (x,y) (Lifegame size board)])

isAlive :: (Int, Int) -> Lifegame -> Bool
isAlive (x, y) (Lifegame size board)
  | nowAlive && numAlive /= 2 && numAlive /= 3  = False
  | not nowAlive && numAlive == 3               = True
  | otherwise = nowAlive
  where numAlive = countAlive (x,y) (Lifegame size board)
        nowAlive = (x,y) `elem` board

countAlive :: (Int, Int) -> Lifegame -> Int
countAlive (x,y) (Lifegame size board) = length $ filter (flip elem [(xx,yy)|xx <- [x-1..x+1], yy <- [y-1..y+1], (xx,yy)/=(x,y)]) board

-- 以下は出力用関数

putCell :: Lifegame -> (Int, Int) -> Char
putCell (Lifegame size board) (x,y) = if (x,y) `elem` board 
                                          then '*'
                                          else ' '

printRow :: Lifegame -> Int -> IO ()
printRow (Lifegame size board) row = print $ map (putCell (Lifegame size board)) [(row,y)|y <- [0..size-1]] 

printBoard :: Lifegame -> IO ()
printBoard (Lifegame size board) = mapM_ (printRow (Lifegame size board)) [x|x <- [0..size-1]]

runLifegame :: Lifegame -> IO String
runLifegame lifegame = do
    printBoard lifegame
    getLine
    runLifegame $ newGeneration lifegame

init_state :: Lifegame
init_state = Lifegame 10 [(1,1), (1,2), (1,3)]

main = do
    runLifegame init_state

File: elixir/lib/lifegame.ex
defmodule LifeGame do
  defstruct size: 0, board: []

  def newGeneration(lifegame = %LifeGame{}) do
    %LifeGame{size: lifegame.size,
              board: (for x <- 0..lifegame.size-1, y <- 0..lifegame.size-1, isAlive(lifegame, x, y), do: {x,y})}
  end

  def countAlive(lifegame = %LifeGame{}, x, y) do
    neigh = for xx <- x-1..x+1, yy <- y-1..y+1, !(xx==x) or !(yy==y), do: {xx, yy}
    lifegame.board
    |> Enum.filter(fn elm -> Enum.member?(neigh, elm) end)
    |> Kernel.length
  end

  def isAlive(lifegame = %LifeGame{}, x, y) do
    numAlive = countAlive(lifegame, x, y)
    nowAlive = Enum.member?(lifegame.board, {x, y})
    cond do
      nowAlive && numAlive != 2 && numAlive != 3
        -> false
      !nowAlive && numAlive == 3
        -> true
      true
        -> nowAlive
    end
  end

  def putCell(lifegame = %LifeGame{}, x, y) do
    if {x, y} in lifegame.board do
      '*'
    else
      ' '
    end
  end

  def printRow(lifegame = %LifeGame{}, row) do
    IO.puts(for y <- 0..lifegame.size, do: putCell(lifegame, row, y))
  end

  def printBoard(lifegame = %LifeGame{}) do
    for x <- 0..lifegame.size, do: printRow(lifegame, x)
  end

  def runLifegame(lifegame = %LifeGame{}) do
    IEx.Helpers.clear()
    printBoard(lifegame)
    IO.puts("")
    :timer.sleep(100)
    runLifegame(newGeneration(lifegame))
  end

  def blinker do
    %LifeGame{size: 5, board: [{1,1}, {1,2}, {1,3}]}
  end

  def pulsar do
    %LifeGame{size: 17, board: [{7,8}, {7,9}, {7,10}, {9,9}, {11,8}, {11,9}, {11,10}]}
  end

  def main do
    runLifegame(pulsar())
  end

end
